In [42]:
import pandas as pd

s1 = pd.read_csv(
    "../student_resource/dataset/train/train_source1.tsv",
    sep="\t"
)

s2 = pd.read_csv(
    "../student_resource/dataset/train/train_source2.tsv",
    sep="\t"
)

s3 = pd.read_csv(
    "../student_resource/dataset/train/train_source3.tsv",
    sep="\t"
)

ground_truth = pd.read_csv(
    "../student_resource/dataset/train/train_ground_truth.tsv",
    sep="\t"
)

In [43]:
print("Source 1:", s1.shape)
print("Source 2:", s2.shape)
print("Source 3:", s3.shape)
print("Ground Truth:", ground_truth.shape)

Source 1: (2206821, 4)
Source 2: (5034616, 4)
Source 3: (5285603, 4)
Ground Truth: (2206821, 2)


In [44]:
print("S1 columns:")
print(s1.columns.tolist())

print("\nS2 columns:")
print(s2.columns.tolist())

print("\nS3 columns:")
print(s3.columns.tolist())

print("\nGround Truth columns:")
print(ground_truth.columns.tolist())

S1 columns:
['entity_id', 'business_name', 'business_address', 'country']

S2 columns:
['entity_id', 'business_name', 'business_address', 'country']

S3 columns:
['entity_id', 'business_name', 'business_address', 'country']

Ground Truth columns:
['source1_entity_id', 'matched_entity_ids']


In [45]:
s1.head()

,entity_id,business_name,business_address,country
0,S1-925783039,Orelee's Barbershop,"1795 Westchester Drive, High Point, NC",US
1,S1-773889195,Prime Money,"17560 Ellis Road, Tahlequah, OK",US
2,S1-377745466,B+ Retail Inc,"1712 Montebello Avenue, Phoenix, AZ",US
3,S1-133037285,Christ Chapel,"2100 Cameron Drive, Unit APARTMENT G, Dundalk, MD",US
4,S1-755362802,Prabhav Business Center,"797, Lake Town Block A, Kolkata, Howrah, West ...",India


In [46]:
s2.head()

,entity_id,business_name,business_address,country
0,S2-166376419,राम मार्केटिंग प्राइवेट लिमिटेड,"KH NO. -570/13, NEW DELHI, WEST DELHI, Delhi",India
1,S2-764573417,-- Holloway Peak Inc Seafood,"105 ELM ST, MORGANTON, NC",US
2,S2-639257739,आदित्य प्रॉपर्टीज एलएलपी,"G-3/571, GULMOHAR COLONY, BHOPAL, Madhya Pradesh",India
3,S2-163963287,Summit Inc,"GREENSBORO, NC, 19 1/2 STARDUST TRAIL",US
4,S2-49942811,Delta Tetlecommunication Inc,"914 PIERPONT AVE, CLEVELAND, OH",US


In [47]:
s3.head()

,entity_id,business_name,business_address,country
0,S3-202863386,wilfordhancock.com,"Mack Rd, Haltom City, Texas",US
1,S3-859268022,International South Consultants Private Ltd,NaN,India
2,S3-22467283,LLC Moncada Léarning Center,"5780 Fawn Ct, Fort Worth, Texas",US
3,S3-671162755,Moyna's Coffee,"1 Ivanhoe Ave, PO Box 6009, Cincinnati, Ohio",US
4,S3-960981775,Pvt. EFS Print Ventures Ltd.,"Door No 183, 41St Cross, 22Nd Main 9Th Block J...",India


In [48]:
ground_truth.head()

,source1_entity_id,matched_entity_ids
0,S1-965667,"S2-681193310,S2-743505751,S3-775321672,S3-1129..."
1,S1-55344266,"S2-249013014,S2-197070651,S3-478195123,S3-3843..."
2,S1-343815751,"S2-790675320,S2-479876582,S3-878454467"
3,S1-656753428,"S2-153058913,S2-24659151,S3-679606215"
4,S1-102811957,"S2-478959098,S2-553508714,S2-625774905,S3-7280..."


In [49]:
print("Unique S1 IDs:", s1["entity_id"].nunique())
print("Unique S2 IDs:", s2["entity_id"].nunique())
print("Unique S3 IDs:", s3["entity_id"].nunique())
print("Unique Ground Truth S1 IDs:", ground_truth["source1_entity_id"].nunique())

Unique S1 IDs: 2206821
Unique S2 IDs: 5034616
Unique S3 IDs: 5285603
Unique Ground Truth S1 IDs: 2206821


- Every Source 1 record has a corresponding Ground Truth row.
- entity_id is unique within each source.

In [50]:
print(ground_truth.head(10).to_string(index=False))

source1_entity_id                                                            matched_entity_ids
        S1-965667               S2-681193310,S2-743505751,S3-775321672,S3-11291185,S3-860443364
      S1-55344266                           S2-249013014,S2-197070651,S3-478195123,S3-384364074
     S1-343815751                                        S2-790675320,S2-479876582,S3-878454467
     S1-656753428                                         S2-153058913,S2-24659151,S3-679606215
     S1-102811957 S2-478959098,S2-553508714,S2-625774905,S3-728090388,S3-928796641,S3-449308785
      S1-18727616                           S2-755677256,S3-187831601,S3-641489370,S3-476250621
     S1-318373630                                                     S2-660036492,S3-804600254
      S1-86989137                                                     S3-274817120,S3-312496301
      S1-29845983                                                     S2-648035184,S3-588502663
     S1-789009573                       

In [51]:
print("Empty matches:", 
      ground_truth["matched_entity_ids"].isna().sum())

print("Empty strings:", 
      (ground_truth["matched_entity_ids"].fillna("") == "").sum())

Empty matches: 123247
Empty strings: 123247


In [52]:
print(
    ground_truth["matched_entity_ids"]
    .fillna("")
    .apply(lambda x: 0 if x == "" else len(x.split(",")))
    .value_counts()
    .sort_index()
    .head(20)
)

matched_entity_ids
0     123247
1     119157
2     375212
3     530841
4     484115
5     321957
6     164868
7      63968
8      18680
9       4205
10       534
11        37
Name: count, dtype: int64


# Ground Truth tells us

- `2,206,821` Source 1 entities:

- `123,247` have no matches → about `5.59`%
- `2,083,574` have at least one match → about `94.41`%
- The number of matches for one S1 entity ranges from `0 to 11`.

In [53]:
print("S1 countries:")
print(s1["country"].value_counts(dropna=False))

print("\nS2 countries:")
print(s2["country"].value_counts(dropna=False))

print("\nS3 countries:")
print(s3["country"].value_counts(dropna=False))

S1 countries:
country
US       1323633
India     883188
Name: count, dtype: int64

S2 countries:
country
US       3016817
India    2017799
Name: count, dtype: int64

S3 countries:
country
US       3170056
India    2115547
Name: count, dtype: int64


In [54]:
print("\nMissing country values:")
print("S1:", s1["country"].isna().sum())
print("S2:", s2["country"].isna().sum())
print("S3:", s3["country"].isna().sum())



Missing country values:
S1: 0
S2: 0
S3: 0


- No country values are missing.
- All training records belong to either `US or India`.
- Country can be useful as an initial blocking constraint.

In [55]:
print("S1 unique business names:", s1["business_name"].nunique())
print("S1 total rows:", len(s1))

print("S1 duplicate business names:",
      s1["business_name"].duplicated().sum())

S1 unique business names: 1539229
S1 total rows: 2206821
S1 duplicate business names: 667592


In [56]:
name_counts = s1["business_name"].value_counts()

print("Most frequent business names:")
print(name_counts.head(20))

Most frequent business names:
business_name
Primary Care Group           253
Ear Nose & Throat Group      251
Pediatric Group              222
Womens Health Group          220
Physical Therapy Group       218
Pediatric Dental Group       216
Behavioral Health Group      215
Chiropractic Group           209
Orthopedic Group             208
Eye Group                    207
Dental Group                 205
Family Group                 204
Dermatology Group            202
Vision Group                 199
Urgent Care Group            198
Pediatric Dentistry Group    198
Internal Medicine Group      197
Foot & Ankle Group           191
Oncology Group               188
Urology Group                185
Name: count, dtype: int64


# S1 has:

- 2,206,821 total records
- 1,539,229 unique business names
- 667,592 records whose business name is duplicated
- The most frequent name appears 253 times

In [57]:
print("S1 unique addresses:", s1["business_address"].nunique())
print("S1 total rows:", len(s1))

print(
    "S1 duplicate addresses:",
    s1["business_address"].duplicated().sum()
)

S1 unique addresses: 2130606
S1 total rows: 2206821
S1 duplicate addresses: 76215


In [58]:
address_counts = s1["business_address"].value_counts()

print("Most frequent addresses:")
print(address_counts.head(20))

Most frequent addresses:
business_address
108 Norle Street, College Twp, PA                                                                                                                                                   14
218 Phillips Street, Storm Lake, IA                                                                                                                                                 13
43 Franklin Street, Delaware, OH                                                                                                                                                    13
1921 Willow Park Drive, Fort Worth, TX                                                                                                                                              13
104 Roadrunner Circle, Elephant Butte, NM                                                                                                                                           13
55, Lavkush Vihar, Singhpur, Kachhar Bithoo

# S1 has:

- 2,206,821 total addresses
- 2,130,606 unique addresses
- 76,215 duplicate-address records
- The most repeated address appears 14 times

### So addresses are much more unique than business names.

In [59]:
for name, df in [("S1", s1), ("S2", s2), ("S3", s3)]:
    print(f"\n{name}")
    print(df[["business_name", "business_address", "country"]].isna().sum())


S1
business_name       0
business_address    0
country             0
dtype: int64

S2
business_name            2
business_address    168967
country                  0
dtype: int64

S3
business_name           13
business_address    175916
country                  0
dtype: int64


In [60]:
for name, df in [("S1", s1), ("S2", s2), ("S3", s3)]:
    print(f"\n{name}")
    for col in ["business_name", "business_address", "country"]:
        print(
            col,
            "empty:",
            (df[col].fillna("").str.strip() == "").sum()
        )


S1
business_name empty: 0
business_address empty: 0
country empty: 0

S2
business_name empty: 2
business_address empty: 168967
country empty: 0

S3
business_name empty: 13
business_address empty: 175916
country empty: 0


# Source 1 is complete, but S2 and S3 contain substantial missing addresses:

- S2: 168,967 missing addresses
- S3: 175,916 missing addresses

### So our matcher cannot depend on address being present.

In [61]:
print("S1:")
display(s1[["business_name", "business_address", "country"]].sample(10, random_state=42))

print("\nS2:")
display(s2[["business_name", "business_address", "country"]].sample(10, random_state=42))

print("\nS3:")
display(s3[["business_name", "business_address", "country"]].sample(10, random_state=42))

S1:


,business_name,business_address,country
2195840,Pediatric Medicine PLLC,"4850 20, Otisco, NY",US
1487051,Fetech National Twin,"19034 Woodburn Road, Woodburn, IN",US
1878277,General Design Innovations LLC,"7241 Osage Avenue, Mesa, AZ",US
993966,Construction Ideaz Papers Private Limited,"Building No.4/606, Prabhul Cottage Karimbalur,...",India
565428,Penaloza and Bittle First Inc.,"4255 Charleswood Avenue, Memphis, TN",US
1645753,National Investments LLC,"1641 Virginia Lane, Hueytown, AL",US
1165724,Reus Idaho Company,"Sanford, ME, 31 Guillemette Street",US
363842,Creative Projects Limited,"12Thfloor, C, 1204, Roya Oasis, Jankalyan Naga...",India
962445,Goar Stone LLC,"2105 Wood Ridge Cove, Cedar Park, TX",US
964820,Lakshmi Consultants Private Limited,"A-301, New Sai Dham Chsl, Ramdev Park Road, Th...",India



S2:


,business_name,business_address,country
2641297,Harbor Center,"6309 EVANGELINE TRAIL, AUSTIN, TX",US
1856454,STEWARD ACE PREMIER,"1029 MAPLE HILL RD, LEBANON, TN",US
371269,Klapper and Lott,"368 VINE STREET, TOOELE, UT",US
793182,EAR NOSE & THROAT CARE,"YORK RD, LUTHERVILLE, MD",US
4159519,Shield Inrfabuild Private Limited,"206 FLAT NO. LG 1 KH NO. 514, 516 SHYAM BHAWAN...",India
2290871,#holtline,"HOUSTON, 6119 DARLINGHURST DRIVE, TX",US
2479261,Cancer [Services],NaN,US
606462,Smt Hermes Private Limited Center,"Delhi, HOUSE NO 138, 2ND FLOOR, BLK C, PKT 2 D...",India
3805434,AMARDEEP PTER PRIVATE LIMITED,"332NEAR TELCO SERVICE STATION RANGPURI, NEW DE...",India
2960649,Impex-+-Brothers,"FLAT B 1801, PL-15&17, SECTOR 7, PATEL HERITAG...",India



S3:


,business_name,business_address,country
726818,Dr Apex Pvt Ltd Partners,"75 Tulshibaugwale Colonysahakarnagar, Pune, MH",India
1045360,Shiva Mines Pvt Ltd,"F-17, Sector 22, Noida, Gautam Buddha Nagar, उ...",India
4785624,Vanguard Métropolitan Sunshine,"3518-B Crestview Ln, Catoosa, Oklahoma",US
2503843,Vanguard Apoge-Inc,"2150 350, Greencastle, Indiana",US
3149825,रियल एग्रो प्राइवेट लिमिटेड,"H.no 829 D-225vivek Vihar, Delhi, दिल्ली",India
2749542,Irinovi,"324 Laramie Ave, Chicago, Illinois",US
3399114,Jarleon Worldwide Llc,"353 Spruce Pine Road, Abingdon, Maryland",US
1611580,Sfoasseeg L.L.C.,"1161d Brookside Ave, Evansdale, Iowa",US
3556867,Brixhalo t/a Creative Enterprises,"6, Madurai Road, Trichy, Tamil Nadu",India
4340188,Nachiket Solutions Limited,"138 / 6, Kiran Complex Zone - Ii, M. P. Nagar ...",India


# The samples show several types of variation we need to handle:

- Different capitalization: `STEWART ACE PREMIER` vs normal title case.
- Abbreviations such as `RD`, `STREET`, `FLAT NO.`, etc.
- Legal/business suffixes such as `Pvt Ltd`, `Private Limited`, `LLC`, Inc.
- Punctuation and unusual characters.
- Indian addresses can contain multiple formats and scripts.
- Some addresses are missing entirely, as we already measured.

However, these are random records, not confirmed matches. So we cannot yet say which differences actually correspond to the same business.

In [62]:
# Select 5 S1 entities that have at least one known match
gt_sample = ground_truth[
    ground_truth["matched_entity_ids"].fillna("").str.strip() != ""
].sample(5, random_state=42)

for _, row in gt_sample.iterrows():

    s1_id = row["source1_entity_id"]
    matched_ids = row["matched_entity_ids"].split(",")

    s1_row = s1[s1["entity_id"] == s1_id]

    print("\n" + "=" * 100)
    print("SOURCE 1")
    print(s1_row[["entity_id", "business_name", "business_address", "country"]].to_string(index=False))

    print("\nMATCHED RECORDS")

    # Check Source 2
    s2_matches = s2[s2["entity_id"].isin(matched_ids)]

    if len(s2_matches) > 0:
        print("\nSource 2:")
        print(
            s2_matches[
                ["entity_id", "business_name", "business_address", "country"]
            ].to_string(index=False)
        )

    # Check Source 3
    s3_matches = s3[s3["entity_id"].isin(matched_ids)]

    if len(s3_matches) > 0:
        print("\nSource 3:")
        print(
            s3_matches[
                ["entity_id", "business_name", "business_address", "country"]
            ].to_string(index=False)
        )


SOURCE 1
   entity_id       business_name             business_address country
S1-777210964 Davis Family Office 88 Olive Circle, Lebanon, TN      US

MATCHED RECORDS

Source 2:
   entity_id       business_name          business_address country
S2-407524980 DAVIS FAMILY OFFICE    OLIVE CIR, LEBANON, TN      US
 S2-12029274  Davis Family Offie 88 OLIVE CIR, LEBANON, TN      US

Source 3:
   entity_id         business_name                 business_address country
S3-970714334 Davis Family (Office) 88 Olive Cir, Lebanon, Tennessee      US
S3-712922821          Davis Family Lebanon, Tennessee, 88 Olive Cir      US

SOURCE 1
   entity_id       business_name                                                                           business_address country
S1-190133982 MS Consultancy Corp Shymala Appts 1St Floor Flat No. 4 Opp Ratna Hospital Sb Road In Haveli, Pune, Maharashtra   India

MATCHED RECORDS

Source 2:
   entity_id    business_name                                                   

In [63]:
for name, df in [("S1", s1), ("S2", s2), ("S3", s3)]:

    print("=" * 60)
    print(name)

    for col in ["business_name", "business_address", "country"]:
        print(f"\n{col}")
        print("Rows:", len(df))
        print("Missing:", df[col].isna().sum())
        print("Unique:", df[col].nunique())
        print("Duplicates:", df[col].duplicated().sum())

S1

business_name
Rows: 2206821
Missing: 0
Unique: 1539229
Duplicates: 667592

business_address
Rows: 2206821
Missing: 0
Unique: 2130606
Duplicates: 76215

country
Rows: 2206821
Missing: 0
Unique: 2
Duplicates: 2206819
S2

business_name
Rows: 5034616
Missing: 2
Unique: 4402008
Duplicates: 632607

business_address
Rows: 5034616
Missing: 168967
Unique: 4337261
Duplicates: 697354

country
Rows: 5034616
Missing: 0
Unique: 2
Duplicates: 5034614
S3

business_name
Rows: 5285603
Missing: 13
Unique: 4651608
Duplicates: 633994

business_address
Rows: 5285603
Missing: 175916
Unique: 4632764
Duplicates: 652838

country
Rows: 5285603
Missing: 0
Unique: 2
Duplicates: 5285601


- country has 0 missing values in all three datasets.
- S2 and S3 have substantial address missingness.
- Business names also contain duplicates across all three sources.
- We should retain the raw columns and create separate processed columns rather than overwriting the original data.

## Objective

Design a preprocessing strategy for the business entity resolution datasets before creating ML-ready features.

The preprocessing should:

- Preserve the original raw data.
- Reduce formatting-based differences.
- Handle missing values safely.
- Keep preprocessing consistent across Source 1, Source 2, and Source 3.
- Avoid aggressive transformations that could create false matches.

---

## 1. Business Name

| Problem | Preprocessing | Reason |
|---|---|---|
| Upper/lowercase differences | Convert to lowercase | Make equivalent text comparable |
| Punctuation differences | Remove punctuation | Reduce formatting variation |
| Extra whitespace | Normalize whitespace | Ensure consistent spacing |
| Unicode variations | Unicode normalization | Handle different Unicode representations |
| Missing values | Fill with empty string + missing indicator | Prevent processing errors while preserving missingness |
| Abbreviations | Do not automatically expand | Avoid incorrect transformations |
| Typos | Do not automatically correct | Avoid introducing false matches |
| Legal suffixes | Do not automatically remove | Preserve potentially useful information |

---

## 2. Business Address

| Problem | Preprocessing | Reason |
|---|---|---|
| Upper/lowercase differences | Convert to lowercase | Make addresses comparable |
| Punctuation differences | Remove punctuation | Reduce formatting variation |
| Extra whitespace | Normalize whitespace | Ensure consistent formatting |
| Unicode variations | Unicode normalization | Handle multilingual/Unicode representations |
| Missing addresses | Fill with empty string + missing indicator | Preserve missingness information |
| Address abbreviations | Do not automatically expand initially | Avoid incorrect transformations |
| Address components in different order | Do not reorder initially | Preserve original information |

---

## 3. Country

| Problem | Preprocessing | Reason |
|---|---|---|
| Case differences | Convert to lowercase | Consistent categorical representation |
| Extra whitespace | Strip whitespace | Remove formatting differences |
| Missing values | No imputation required | No missing country values were found |

---

## 4. Transformations We Will Not Apply Automatically

The initial preprocessing will **not** automatically:

- Correct spelling mistakes
- Expand abbreviations such as `Cir` → `Circle`
- Expand abbreviations such as `Rd` → `Road`
- Remove legal suffixes such as `LLC`, `Inc`, or `Pvt Ltd`
- Translate business names or addresses
- Create manual aliases
- Reorder address components

These transformations will only be considered later if analysis of the training data provides sufficient evidence.

---

## 5. Data Preservation Principle

The original columns will remain unchanged.

New processed columns will be created:

```text
business_name
business_name_clean

business_address
business_address_clean

country
country_clean

## Basic Text Preprocessing

In [64]:
import re
import unicodedata

def normalize_text(text):
    if pd.isna(text):  # Check for missing values
        return ""

    text = str(text).lower()
    text = unicodedata.normalize("NFKC", text) # Different Unicode representations → standardized representation
    text = re.sub(r"[^\w\s]", " ", text) # Remove punctuation and special characters
    text = re.sub(r"\s+", " ", text).strip() # remove extra space

    return text

In [65]:
s1["business_name_clean"] = s1["business_name"].map(normalize_text)

s1["business_address_clean"] = s1["business_address"].map(normalize_text)

s1["country_clean"] = (
    s1["country"]
    .str.lower()
    .str.strip()
)

In [66]:
display(
    s1[
        [
            "business_name",
            "business_name_clean",
            "business_address",
            "business_address_clean",
            "country",
            "country_clean"
        ]
    ].head(10)
)

,business_name,business_name_clean,business_address,business_address_clean,country,country_clean
0,Orelee's Barbershop,orelee s barbershop,"1795 Westchester Drive, High Point, NC",1795 westchester drive high point nc,US,us
1,Prime Money,prime money,"17560 Ellis Road, Tahlequah, OK",17560 ellis road tahlequah ok,US,us
2,B+ Retail Inc,b retail inc,"1712 Montebello Avenue, Phoenix, AZ",1712 montebello avenue phoenix az,US,us
3,Christ Chapel,christ chapel,"2100 Cameron Drive, Unit APARTMENT G, Dundalk, MD",2100 cameron drive unit apartment g dundalk md,US,us
4,Prabhav Business Center,prabhav business center,"797, Lake Town Block A, Kolkata, Howrah, West ...",797 lake town block a kolkata howrah west bengal,India,india
5,Custom Wealth Services LLC,custom wealth services llc,"OH, Columbus, 5559 Orville Avenue",oh columbus 5559 orville avenue,US,us
6,Consulting Nyasa Nursing Private Limited,consulting nyasa nursing private limited,"2505, Tower 1, Oakwood, Runwal Greens, Mulund ...",2505 tower 1 oakwood runwal greens mulund gore...,India,india
7,Nexus Anchor Rain,nexus anchor rain,"1111 Church Street, Unit 2007, Nashville, TN",1111 church street unit 2007 nashville tn,US,us
8,Moore Bitwise Inc,moore bitwise inc,"337 Oakland Avenue, Michigan City, IN",337 oakland avenue michigan city in,US,us
9,Dermatology Green Medicine,dermatology green medicine,"294 Meadowcreek Drive, Unit Unit 2, Village Of...",294 meadowcreek drive unit unit 2 village of p...,US,us


## Validate the preprocessing

In [67]:
print("S1 original vs cleaned")

print("Original business_name missing:",
      s1["business_name"].isna().sum())

print("Clean business_name missing:",
      s1["business_name_clean"].isna().sum())

print("Original address missing:",
      s1["business_address"].isna().sum())

print("Clean address missing:",
      s1["business_address_clean"].isna().sum())

S1 original vs cleaned
Original business_name missing: 0
Clean business_name missing: 0
Original address missing: 0
Clean address missing: 0


In [68]:
print("\nEmpty cleaned values:")

print(
    "business_name_clean:",
    (s1["business_name_clean"] == "").sum()
)

print(
    "business_address_clean:",
    (s1["business_address_clean"] == "").sum()
)

print(
    "country_clean:",
    (s1["country_clean"] == "").sum()
)


Empty cleaned values:
business_name_clean: 0
business_address_clean: 0
country_clean: 0


In [69]:
print("\nUnique values")

print(
    "business_name:",
    s1["business_name"].nunique(),
    "→",
    s1["business_name_clean"].nunique()
)

print(
    "business_address:",
    s1["business_address"].nunique(),
    "→",
    s1["business_address_clean"].nunique()
)

print(
    "country:",
    s1["country"].nunique(),
    "→",
    s1["country_clean"].nunique()
)


Unique values
business_name: 1539229 → 1520684
business_address: 2130606 → 2130172
country: 2 → 2


## Apply preprocessing to S2 and S3

In [70]:
# S2
s2["business_name_clean"] = s2["business_name"].map(normalize_text)

s2["business_address_clean"] = s2["business_address"].map(normalize_text)

s2["country_clean"] = (
    s2["country"]
    .str.lower()
    .str.strip()
)

In [71]:
# S3
s3["business_name_clean"] = s3["business_name"].map(normalize_text)

s3["business_address_clean"] = s3["business_address"].map(normalize_text)

s3["country_clean"] = (
    s3["country"]
    .str.lower()
    .str.strip()
)

In [72]:
for name, df in [("S1", s1), ("S2", s2), ("S3", s3)]:
    print("=" * 50)
    print(name)

    print(
        "Empty business names:",
        (df["business_name_clean"] == "").sum()
    )

    print(
        "Empty addresses:",
        (df["business_address_clean"] == "").sum()
    )

    print(
        "Empty countries:",
        (df["country_clean"] == "").sum()
    )

S1
Empty business names: 0
Empty addresses: 0
Empty countries: 0
S2
Empty business names: 2
Empty addresses: 168967
Empty countries: 0
S3
Empty business names: 13
Empty addresses: 175916
Empty countries: 0


# Missing-Value Indicators

In [73]:
for df in [s1, s2, s3]:

    df["business_name_missing"] = (
        df["business_name"].isna()
    ).astype("int8")

    df["business_address_missing"] = (
        df["business_address"].isna()
    ).astype("int8")

In [74]:
for name, df in [("S1", s1), ("S2", s2), ("S3", s3)]:

    print("=" * 50)
    print(name)

    print(
        "Business name missing:",
        df["business_name_missing"].sum()
    )

    print(
        "Business address missing:",
        df["business_address_missing"].sum()
    )

S1
Business name missing: 0
Business address missing: 0
S2
Business name missing: 2
Business address missing: 168967
S3
Business name missing: 13
Business address missing: 175916


# Analyze Duplicate Records

In [75]:
for name, df in [("S1", s1), ("S2", s2), ("S3", s3)]:

    print("=" * 60)
    print(name)

    print(
        "Exact duplicate rows:",
        df.duplicated(
            subset=["business_name", "business_address", "country"]
        ).sum()
    )

    print(
        "Duplicate name + address:",
        df.duplicated(
            subset=[
                "business_name_clean",
                "business_address_clean",
                "country_clean"
            ]
        ).sum()
    )

S1
Exact duplicate rows: 0
Duplicate name + address: 0
S2
Exact duplicate rows: 25873
Duplicate name + address: 67919
S3
Exact duplicate rows: 18860
Duplicate name + address: 50537


For this entity-resolution challenge, multiple records can legitimately refer to the same real-world entity, and the Ground Truth allows one S1 entity to have multiple matches.

So our preprocessing report should record these duplicates, but preserve the rows for the ML pipeline.

# Quantify Duplicate Groups

In [76]:
for name, df in [("S1", s1), ("S2", s2), ("S3", s3)]:

    group_counts = (
        df.groupby(
            [
                "business_name_clean",
                "business_address_clean",
                "country_clean"
            ],
            dropna=False
        )
        .size()
    )

    duplicate_groups = group_counts[group_counts > 1]

    print("=" * 60)
    print(name)
    print("Duplicate groups:", len(duplicate_groups))
    print("Records inside duplicate groups:", duplicate_groups.sum())
    print("Largest duplicate group:", duplicate_groups.max())

S1
Duplicate groups: 0
Records inside duplicate groups: 0
Largest duplicate group: nan
S2
Duplicate groups: 65189
Records inside duplicate groups: 133108
Largest duplicate group: 5
S3
Duplicate groups: 48678
Records inside duplicate groups: 99215
Largest duplicate group: 4


- S1 has no duplicate cleaned (name, address, country) combinations.
- S2 has 65,189 duplicate groups.
- S3 has 48,678 duplicate groups.
- No duplicate group is extremely large: maximum 5 in S2 and 4 in S3.
- These records should not be automatically removed, because they may represent multiple source - records referring to the same real-world business.

# Data Quality Summary

In [77]:
quality_report = pd.DataFrame({
    "dataset": ["S1", "S2", "S3"],
    "rows": [
        len(s1),
        len(s2),
        len(s3)
    ],
    "missing_business_name": [
        s1["business_name_missing"].sum(),
        s2["business_name_missing"].sum(),
        s3["business_name_missing"].sum()
    ],
    "missing_business_address": [
        s1["business_address_missing"].sum(),
        s2["business_address_missing"].sum(),
        s3["business_address_missing"].sum()
    ],
    "duplicate_exact_rows": [
        s1.duplicated(
            subset=["business_name", "business_address", "country"]
        ).sum(),
        s2.duplicated(
            subset=["business_name", "business_address", "country"]
        ).sum(),
        s3.duplicated(
            subset=["business_name", "business_address", "country"]
        ).sum()
    ],
    "duplicate_clean_name_address": [
        s1.duplicated(
            subset=[
                "business_name_clean",
                "business_address_clean",
                "country_clean"
            ]
        ).sum(),
        s2.duplicated(
            subset=[
                "business_name_clean",
                "business_address_clean",
                "country_clean"
            ]
        ).sum(),
        s3.duplicated(
            subset=[
                "business_name_clean",
                "business_address_clean",
                "country_clean"
            ]
        ).sum()
    ]
})

quality_report

,dataset,rows,missing_business_name,missing_business_address,duplicate_exact_rows,duplicate_clean_name_address
0,S1,2206821,0,0,0,0
1,S2,5034616,2,168967,25873,67919
2,S3,5285603,13,175916,18860,50537


# Analyze Name Normalization Impact


In [78]:
name_normalization_report = pd.DataFrame({
    "dataset": ["S1", "S2", "S3"],
    "original_unique_names": [
        s1["business_name"].nunique(),
        s2["business_name"].nunique(),
        s3["business_name"].nunique()
    ],
    "clean_unique_names": [
        s1["business_name_clean"].nunique(),
        s2["business_name_clean"].nunique(),
        s3["business_name_clean"].nunique()
    ],
})

name_normalization_report["reduced_unique_names"] = (
    name_normalization_report["original_unique_names"]
    - name_normalization_report["clean_unique_names"]
)

name_normalization_report

,dataset,original_unique_names,clean_unique_names,reduced_unique_names
0,S1,1539229,1520684,18545
1,S2,4402008,4025057,376951
2,S3,4651608,4280400,371208


In [79]:
name_normalization_report["reduction_percent"] = (
    name_normalization_report["reduced_unique_names"]
    / name_normalization_report["original_unique_names"]
    * 100
)

name_normalization_report

,dataset,original_unique_names,clean_unique_names,reduced_unique_names,reduction_percent
0,S1,1539229,1520684,18545,1.204824
1,S2,4402008,4025057,376951,8.563160
2,S3,4651608,4280400,371208,7.980208


- basic normalization has a moderate effect on S2/S3 but a much smaller effect on S1.

business_name != business_name_clean

In [80]:
changed_s1 = s1[
    s1["business_name"] != s1["business_name_clean"]
][
    ["business_name", "business_name_clean"]
]

display(
    changed_s1.sample(20, random_state=42)
)


,business_name,business_name_clean
2195840,Pediatric Medicine PLLC,pediatric medicine pllc
1487051,Fetech National Twin,fetech national twin
1878277,General Design Innovations LLC,general design innovations llc
993966,Construction Ideaz Papers Private Limited,construction ideaz papers private limited
565428,Penaloza and Bittle First Inc.,penaloza and bittle first inc
1645753,National Investments LLC,national investments llc
1165724,Reus Idaho Company,reus idaho company
363842,Creative Projects Limited,creative projects limited
962445,Goar Stone LLC,goar stone llc
964820,Lakshmi Consultants Private Limited,lakshmi consultants private limited


In [81]:
changed_s2 = s2[
    s2["business_name"] != s2["business_name_clean"]
][
    ["business_name", "business_name_clean"]
]

display(
    changed_s2.sample(20, random_state=42)
)

,business_name,business_name_clean
2197285,Heartland Program Metro,heartland program metro
4116940,NAVI MUMBAI INFRA PVT LTD.,navi mumbai infra pvt ltd
142297,2871 Church Lane,2871 church lane
4741012,Mumbai Automotives Automotives Pvt. Ltd,mumbai automotives automotives pvt ltd
1741229,Sullivan Holdings Holdings,sullivan holdings holdings
4844095,PLEX FOODS PRIVATE,plex foods private
2392774,"Pediatric Dental Scnue Partners,-LLC",pediatric dental scnue partners llc
1820480,Horizon Express Taernsport,horizon express taernsport
315658,G 7 R Resources Group,g 7 r resources group
4355758,Mount Emanuel,mount emanuel


# Analyze Name Token Structure

In [82]:
def name_token_count(text):
    if not text:
        return 0
    return len(text.split())


for name, df in [("S1", s1), ("S2", s2), ("S3", s3)]:

    token_counts = df["business_name_clean"].map(name_token_count)

    print("=" * 50)
    print(name)

    print("Average tokens:", token_counts.mean())
    print("Minimum tokens:", token_counts.min())
    print("Maximum tokens:", token_counts.max())

    print("\nToken distribution:")
    print(token_counts.value_counts().sort_index().head(15))

S1
Average tokens: 3.5739282887012584
Minimum tokens: 1
Maximum tokens: 19

Token distribution:
business_name_clean
1      14598
2     307227
3     672164
4     910703
5     241933
6      42643
7      10963
8       4115
9       1661
10       530
11       222
12        41
13        12
14         3
15         2
Name: count, dtype: int64
S2
Average tokens: 4.493855539330109
Minimum tokens: 0
Maximum tokens: 30

Token distribution:
business_name_clean
0           2
1      110171
2      830029
3     1270129
4     1432528
5      659026
6      206361
7       69001
8       36936
9       28314
10      28188
11      28998
12      34706
13      44246
14      55614
Name: count, dtype: int64
S3
Average tokens: 4.129734866580029
Minimum tokens: 0
Maximum tokens: 29

Token distribution:
business_name_clean
0          13
1      182018
2      865695
3     1308520
4     1532132
5      727284
6      254147
7      102119
8       58782
9       29995
10      22179
11      18361
12      20684
13      24934
1

- Most business names contain 3–4 tokens.
- S2 and S3 contain a small number of names with 0 tokens, corresponding to the missing names we already identified.
- S2/S3 have considerably longer names than the typical S1 name.
Token count can therefore become a useful ML feature, but we should not use it as a matching rule by itself.

# Basic Name Features

In [83]:
def add_name_features(df):

    df["name_char_length"] = (
        df["business_name_clean"].str.len().astype("int32")
    )

    df["name_token_count"] = (
        df["business_name_clean"]
        .str.split()
        .str.len()
        .astype("int16")
    )

    df["name_unique_token_count"] = (
        df["business_name_clean"]
        .str.split()
        .map(lambda x: len(set(x)))
        .astype("int16")
    )

    return df

In [84]:
s1 = add_name_features(s1)
s2 = add_name_features(s2)
s3 = add_name_features(s3)

In [85]:
display(
    s1[
        [
            "business_name_clean",
            "name_char_length",
            "name_token_count",
            "name_unique_token_count"
        ]
    ].head(10)
)

,business_name_clean,name_char_length,name_token_count,name_unique_token_count
0,orelee s barbershop,19,3,3
1,prime money,11,2,2
2,b retail inc,12,3,3
3,christ chapel,13,2,2
4,prabhav business center,23,3,3
5,custom wealth services llc,26,4,4
6,consulting nyasa nursing private limited,40,5,5
7,nexus anchor rain,17,3,3
8,moore bitwise inc,17,3,3
9,dermatology green medicine,26,3,3


# Basic Address Features

In [86]:
def add_address_features(df):

    df["address_char_length"] = (
        df["business_address_clean"].str.len().astype("int32")
    )

    df["address_token_count"] = (
        df["business_address_clean"]
        .str.split()
        .str.len()
        .astype("int16")
    )

    df["address_digit_count"] = (
        df["business_address_clean"]
        .str.count(r"\d")
        .astype("int16")
    )

    df["address_unique_token_count"] = (
        df["business_address_clean"]
        .str.split()
        .map(lambda x: len(set(x)))
        .astype("int16")
    )

    return df

In [87]:
s1 = add_address_features(s1)
s2 = add_address_features(s2)
s3 = add_address_features(s3)

In [88]:
display(
    s1[
        [
            "business_address_clean",
            "address_char_length",
            "address_token_count",
            "address_digit_count",
            "address_unique_token_count"
        ]
    ].head(10)
)

,business_address_clean,address_char_length,address_token_count,address_digit_count,address_unique_token_count
0,1795 westchester drive high point nc,36,6,4,6
1,17560 ellis road tahlequah ok,29,5,5,5
2,1712 montebello avenue phoenix az,33,5,4,5
3,2100 cameron drive unit apartment g dundalk md,46,8,4,8
4,797 lake town block a kolkata howrah west bengal,48,9,3,9
5,oh columbus 5559 orville avenue,31,5,4,5
6,2505 tower 1 oakwood runwal greens mulund gore...,113,17,5,17
7,1111 church street unit 2007 nashville tn,41,7,8,7
8,337 oakland avenue michigan city in,35,6,3,6
9,294 meadowcreek drive unit unit 2 village of p...,56,10,4,9


# Validate Address Features Against Missing Data


In [89]:
for name, df in [("S1", s1), ("S2", s2), ("S3", s3)]:

    missing = df["business_address_missing"] == 1

    print("=" * 60)
    print(name)

    print("Missing address rows:", missing.sum())

    if missing.sum() > 0:
        print(
            "Address character length:",
            df.loc[missing, "address_char_length"].unique()
        )

        print(
            "Address token count:",
            df.loc[missing, "address_token_count"].unique()
        )

        print(
            "Address digit count:",
            df.loc[missing, "address_digit_count"].unique()
        )

        print(
            "Address unique token count:",
            df.loc[missing, "address_unique_token_count"].unique()
        )

S1
Missing address rows: 0
S2
Missing address rows: 168967
Address character length: [0]
Address token count: [0]
Address digit count: [0]
Address unique token count: [0]
S3
Missing address rows: 175916
Address character length: [0]
Address token count: [0]
Address digit count: [0]
Address unique token count: [0]


# Final ML Feature Summary

In [90]:
feature_summary = pd.DataFrame({
    "feature": [
        "business_name_clean",
        "business_address_clean",
        "country_clean",
        "business_name_missing",
        "business_address_missing",
        "name_char_length",
        "name_token_count",
        "name_unique_token_count",
        "address_char_length",
        "address_token_count",
        "address_digit_count",
        "address_unique_token_count"
    ],
    "type": [
        "text",
        "text",
        "categorical",
        "binary",
        "binary",
        "numeric",
        "numeric",
        "numeric",
        "numeric",
        "numeric",
        "numeric",
        "numeric"
    ],
    "purpose": [
        "Normalized business name",
        "Normalized business address",
        "Standardized country",
        "Indicates missing business name",
        "Indicates missing business address",
        "Business name length",
        "Number of business name tokens",
        "Number of unique business name tokens",
        "Business address length",
        "Number of address tokens",
        "Number of digits in address",
        "Number of unique address tokens"
    ]
})

feature_summary

,feature,type,purpose
0,business_name_clean,text,Normalized business name
1,business_address_clean,text,Normalized business address
2,country_clean,categorical,Standardized country
3,business_name_missing,binary,Indicates missing business name
4,business_address_missing,binary,Indicates missing business address
5,name_char_length,numeric,Business name length
6,name_token_count,numeric,Number of business name tokens
7,name_unique_token_count,numeric,Number of unique business name tokens
8,address_char_length,numeric,Business address length
9,address_token_count,numeric,Number of address tokens


# Final Feature Quality Check

In [91]:
feature_columns = [
    "business_name_clean",
    "business_address_clean",
    "country_clean",
    "business_name_missing",
    "business_address_missing",
    "name_char_length",
    "name_token_count",
    "name_unique_token_count",
    "address_char_length",
    "address_token_count",
    "address_digit_count",
    "address_unique_token_count"
]

for name, df in [("S1", s1), ("S2", s2), ("S3", s3)]:

    print("=" * 60)
    print(name)

    print("Rows:", len(df))

    print("\nMissing values:")
    print(df[feature_columns].isna().sum())

    print("\nNumeric infinity:")
    numeric_cols = df[feature_columns].select_dtypes(
        include=["number"]
    ).columns

    print(
        df[numeric_cols]
        .isin([float("inf"), float("-inf")])
        .sum()
    )

    print("\nData types:")
    print(df[feature_columns].dtypes)

S1
Rows: 2206821

Missing values:
business_name_clean           0
business_address_clean        0
country_clean                 0
business_name_missing         0
business_address_missing      0
name_char_length              0
name_token_count              0
name_unique_token_count       0
address_char_length           0
address_token_count           0
address_digit_count           0
address_unique_token_count    0
dtype: int64

Numeric infinity:
business_name_missing         0
business_address_missing      0
name_char_length              0
name_token_count              0
name_unique_token_count       0
address_char_length           0
address_token_count           0
address_digit_count           0
address_unique_token_count    0
dtype: int64

Data types:
business_name_clean           object
business_address_clean        object
country_clean                 object
business_name_missing           int8
business_address_missing        int8
name_char_length               int32
name_token_cou

# Data Preprocessing Report

## 1. Objective

The objective of preprocessing is to transform the raw business entity datasets into consistent, clean, and ML-ready data while preserving the original information.

The preprocessing was performed independently on Source 1, Source 2, and Source 3.

---

## 2. Dataset Overview

| Dataset | Rows | Columns |
|---|---:|---:|
| Source 1 | 2,206,821 | 4 + engineered features |
| Source 2 | 5,034,616 | 4 + engineered features |
| Source 3 | 5,285,603 | 4 + engineered features |

Original fields:

- `entity_id`
- `business_name`
- `business_address`
- `country`

---

## 3. Data Quality Findings

### Missing Values

| Dataset | Missing Business Name | Missing Business Address |
|---|---:|---:|
| S1 | 0 | 0 |
| S2 | 2 | 168,967 |
| S3 | 13 | 175,916 |

Country contained no missing values in any source.

---

## 4. Duplicate Analysis

| Dataset | Exact Duplicate Rows | Duplicate Clean Name + Address |
|---|---:|---:|
| S1 | 0 | 0 |
| S2 | 25,873 | 67,919 |
| S3 | 18,860 | 50,537 |

Duplicates were identified and reported but not automatically removed because multiple records may represent the same real-world entity.

---

## 5. Text Normalization

The following normalization was applied to business names and addresses:

1. Convert text to lowercase.
2. Apply Unicode NFKC normalization.
3. Remove punctuation.
4. Normalize consecutive whitespace.
5. Strip leading and trailing whitespace.

The original columns were preserved.

New columns:

- `business_name_clean`
- `business_address_clean`
- `country_clean`

---

## 6. Missing-Value Indicators

Explicit missing-value indicators were created:

- `business_name_missing`
- `business_address_missing`

Missing text values are represented by empty cleaned strings while the indicator columns preserve information about the original missingness.

---

## 7. Name Features

The following features were generated:

- `name_char_length`
- `name_token_count`
- `name_unique_token_count`

These features describe the structure of the normalized business name.

---

## 8. Address Features

The following features were generated:

- `address_char_length`
- `address_token_count`
- `address_digit_count`
- `address_unique_token_count`

These features describe the structure of the normalized business address.

---

## 9. Feature Quality Validation

The generated features were checked for:

- Missing values
- Infinite values
- Appropriate data types
- Preserved row counts

The preprocessing pipeline did not intentionally remove records.

---

## 10. Important Preprocessing Decisions

The following transformations were intentionally NOT applied automatically:

- Spelling correction
- Legal-suffix removal
- Abbreviation expansion
- Manual business-name aliases
- Translation
- Address component reordering

These transformations may affect entity-matching precision and should be evaluated by the ML/matching team before implementation.

---

## 11. ML Handoff

The processed datasets are ready for the next stage of the pipeline.

The ML team can use:

### Text fields

- `business_name_clean`
- `business_address_clean`
- `country_clean`

### Missingness features

- `business_name_missing`
- `business_address_missing`

### Name features

- `name_char_length`
- `name_token_count`
- `name_unique_token_count`

### Address features

- `address_char_length`
- `address_token_count`
- `address_digit_count`
- `address_unique_token_count`

The original raw fields remain available for traceability and debugging.

---

## 12. Handoff Boundary

### Data preprocessing team

Responsible for:

- Data quality analysis
- Cleaning
- Normalization
- Missing-value handling
- Basic feature generation
- Data-quality documentation

### ML / Matching team

Next responsibilities:

- Candidate generation / blocking
- Pairwise similarity features
- Entity-matching model
- Match threshold selection
- Evaluation using the challenge metric
- Final prediction generation

In [103]:
import os

processed_path = "../student_resource/dataset/processed"

os.makedirs(processed_path, exist_ok=True)

s1.to_parquet(
    f"{processed_path}/train_source1_processed.parquet",
    index=False
)

s2.to_parquet(
    f"{processed_path}/train_source2_processed.parquet",
    index=False
)

s3.to_parquet(
    f"{processed_path}/train_source3_processed.parquet",
    index=False
)

print(os.listdir(processed_path))

['train_source1_processed.parquet', 'train_source2_processed.parquet', 'train_source3_processed.parquet']


In [ ]:
import pandas as pd

processed_path = "../student_resource/dataset/processed"

s1_check = pd.read_parquet(
    f"{processed_path}/train_source1_processed.parquet"
)

s2_check = pd.read_parquet(
    f"{processed_path}/train_source2_processed.parquet"
)

s3_check = pd.read_parquet(
    f"{processed_path}/train_source3_processed.parquet"
)

for name, original, processed in [
    ("S1", s1, s1_check),
    ("S2", s2, s2_check),
    ("S3", s3, s3_check)
]:
    print("=" * 50)
    print(name)
    print("Original shape:", original.shape)
    print("Processed shape:", processed.shape)
    print("Rows preserved:", len(original) == len(processed))

S1
Original shape: (2206821, 16)
Processed shape: (2206821, 16)
Rows preserved: True
S2
Original shape: (5034616, 16)
Processed shape: (5034616, 16)
Rows preserved: True
S3
Original shape: (5285603, 16)
Processed shape: (5285603, 16)
Rows preserved: True


In [108]:
s1_check.head()

,entity_id,business_name,business_address,country,business_name_clean,business_address_clean,country_clean,business_name_missing,business_address_missing,name_char_length,name_token_count,name_unique_token_count,address_char_length,address_token_count,address_digit_count,address_unique_token_count
0,S1-925783039,Orelee's Barbershop,"1795 Westchester Drive, High Point, NC",US,orelee s barbershop,1795 westchester drive high point nc,us,0,0,19,3,3,36,6,4,6
1,S1-773889195,Prime Money,"17560 Ellis Road, Tahlequah, OK",US,prime money,17560 ellis road tahlequah ok,us,0,0,11,2,2,29,5,5,5
2,S1-377745466,B+ Retail Inc,"1712 Montebello Avenue, Phoenix, AZ",US,b retail inc,1712 montebello avenue phoenix az,us,0,0,12,3,3,33,5,4,5
3,S1-133037285,Christ Chapel,"2100 Cameron Drive, Unit APARTMENT G, Dundalk, MD",US,christ chapel,2100 cameron drive unit apartment g dundalk md,us,0,0,13,2,2,46,8,4,8
4,S1-755362802,Prabhav Business Center,"797, Lake Town Block A, Kolkata, Howrah, West ...",India,prabhav business center,797 lake town block a kolkata howrah west bengal,india,0,0,23,3,3,48,9,3,9
